File: core/models/sale.py
from dataclasses import dataclass, field
from datetime import datetime
from decimal import Decimal
from typing import List, Optional
from core.models.enums import PaymentType

# Assuming Product model is defined elsewhere or not needed directly for definition
# from core.models.product import Product


@dataclass
class SaleItem:
    # Fields without defaults first
    product_id: int
    quantity: Decimal  # This will be converted to Decimal if float
    unit_price: Decimal  # This will be converted to Decimal if float

    # Fields with defaults next
    id: Optional[int] = None
    sale_id: Optional[int] = None
    product_code: str = ""  # Denormalized for easy display
    product_description: str = ""  # Denormalized for easy display
    product_unit: str = "Unidad"  # Unit of measure for the product

    def __post_init__(self):
        # Ensure quantity and unit_price are always Decimal objects
        if not isinstance(self.quantity, Decimal):
            self.quantity = Decimal(str(self.quantity))
        if not isinstance(self.unit_price, Decimal):
            self.unit_price = Decimal(str(self.unit_price))

    @property
    def subtotal(self) -> Decimal:
        return (self.quantity * self.unit_price).quantize(Decimal("0.01"))

    # You might add product details here if needed, fetched separately or passed during creation


@dataclass
class Sale:
    id: Optional[int] = None
    timestamp: datetime = field(default_factory=datetime.now)
    items: List[SaleItem] = field(default_factory=list)
    customer_id: Optional[int] = None  # Added customer ID
    is_credit_sale: bool = False  # Added credit flag
    user_id: Optional[int] = None  # User who made the sale
    payment_type: Optional["PaymentType"] = (
        None  # e.g., 'Efectivo', 'Tarjeta', 'Crédito'
    )
    # status: str = "COMPLETED" # Example status

    @property
    def total(self) -> Decimal:
        if not self.items:
            return Decimal("0.00")
        return sum(item.subtotal for item in self.items).quantize(Decimal("0.01"))

File: csharp/src/Eleventa.Domain/Entities/Sale.cs
namespace Eleventa.Domain.Entities;

/// <summary>
/// Represents a sale transaction.
/// </summary>
public class Sale
{
    public int Id { get; set; }
    public DateTime Timestamp { get; set; } = DateTime.Now;
    public int? CustomerId { get; set; }
    public int? UserId { get; set; }
    public bool IsCreditSale { get; set; }
    public PaymentType? PaymentType { get; set; }
    public string? Status { get; set; }

    // Navigation properties
    public Customer? Customer { get; set; }
    public User? User { get; set; }
    public ICollection<SaleItem> Items { get; set; } = new List<SaleItem>();

    /// <summary>
    /// Calculates the total amount of the sale.
    /// </summary>
    public decimal Total
    {
        get
        {
            if (Items == null || !Items.Any())
                return 0m;

            return Items.Sum(item => item.Subtotal);
        }
    }

    /// <summary>
    /// Gets the total number of items in the sale.
    /// </summary>
    public int ItemCount => Items?.Count ?? 0;

    /// <summary>
    /// Adds an item to the sale.
    /// </summary>
    public void AddItem(int productId, string productCode, string productDescription,
                       string productUnit, decimal quantity, decimal unitPrice)
    {
        var item = new SaleItem
        {
            ProductId = productId,
            ProductCode = productCode,
            ProductDescription = productDescription,
            ProductUnit = productUnit,
            Quantity = quantity,
            UnitPrice = unitPrice
        };

        Items.Add(item);
    }
}

/// <summary>
/// Represents an item within a sale.
/// </summary>
public class SaleItem
{
    public int Id { get; set; }
    public int? SaleId { get; set; }
    public int ProductId { get; set; }

    // Denormalized fields for display
    public string ProductCode { get; set; } = string.Empty;
    public string ProductDescription { get; set; } = string.Empty;
    public string ProductUnit { get; set; } = "Unidad";

    public decimal Quantity { get; set; }
    public decimal UnitPrice { get; set; }

    // Navigation properties
    public Sale? Sale { get; set; }
    public Product? Product { get; set; }

    /// <summary>
    /// Calculates the subtotal for this line item.
    /// </summary>
    public decimal Subtotal => Math.Round(Quantity * UnitPrice, 2);
}
